src/Revisionable/RevisionableListener.php

```diff
@@ -0,0 +1,390 @@
+
+        foreach ($ea->getScheduledObjectDeletions($uow) as $object) {
+            $this->createRevision(RevisionInterface::ACTION_REMOVE, $object, $ea);
+        }
+    }
+
+    protected function getNamespace(): string
+    {
+        return __NAMESPACE__;
+    }
+
+    /**
+     * Get the {@see RevisionInterface} class name to use when creating revisions for the provided class.
+     *
+     * @param class-string $class
+     *
+     * @return class-string<RevisionInterface<T>>
+     */
+    private function getRevisionClass(RevisionableAdapter $ea, string $class): string
+    {
+        return $this->getConfiguration($ea->getObjectManager(), $class)['revisionClass'] ?? $ea->getDefaultRevisionClass();
+    }
+
+    /**
+     * Retrieve the username to use for the log entry.
+     *
+     * This method will try to fetch a username from the actor provider first, falling back to the {@see $this->username}
+     * property if the provider is not set or does not provide a value.
+     *
+     * @throws UnexpectedValueException if the actor provider provides an unsupported username value
+     *
+     * @return non-empty-string|null
+     */
+    private function getUsername(): ?string
+    {
+        if ($this->actorProvider instanceof ActorProviderInterface) {
+            $actor = $this->actorProvider->getActor();
+
+            if (is_string($actor) || null === $actor) {
+                return $actor;
+            }
+
+            if (method_exists($actor, 'getUserIdentifier')) {
+                return (string) $actor->getUserIdentifier();
+            }
+
+            if (method_exists($actor, 'getUsername')) {
+                return (string) $actor->getUsername();
+            }
+
+            if (method_exists($actor, '__toString')) {
+                return $actor->__toString();
+            }
+
+            throw new UnexpectedValueException(\sprintf('The revisionable extension requires the actor provider to return a string or an object implementing the "getUserIdentifier()", "getUsername()", or "__toString()" methods. "%s" cannot be used as an actor.', get_class($actor)));
+        }
+
+        return $this->username;
+    }
+
+    /**
+     * Provides the changed data for an object to store in a revision.
+     *
+     * @param T                    $object
+     * @param RevisionInterface<T> $revision
+     *
+     * @return array<string, mixed>
+     */
+    private function getObjectChangeSetData(RevisionableAdapter $ea, object $object, RevisionInterface $revision): array
+    {
+        $om = $ea->getObjectManager();
+        $wrapped = AbstractWrapper::wrap($object, $om);
+        $meta = $wrapped->getMetadata();
+        $config = $this->getConfiguration($om, $meta->getName());
+        $uow = $om->getUnitOfWork();
+        $newValues = [];
+
+        foreach ($ea->getObjectChangeSet($uow, $object) as $field => $changes) {
+            if (empty($config['versioned']) || !in_array($field, $config['versioned'], true)) {
```

Comment:
Do you think we can replace all the added conditions using `empty()` with its stricter variants?

I'd like to reduce the work we must do later, narrowing conditions like this in the existing codebase.

Reply:
Yeah, I'll see what I can do with it.